composable_mutex.ml

```diff
@@ -0,0 +1,32 @@
+open Effect
+
+type state = Unlocked | Locked of unit Sched.resumer list
+type t = state Atomic.t
+
+let create () = Atomic.make Unlocked
+
+let lock m =
+  let rec block r =
+    let old = Atomic.get m in
+    match old with
+    | Unlocked ->
+        if Atomic.compare_and_set m old (Locked []) then Some ()
+        else block r (* failed CAS; retry *)
+    | Locked l ->
+        if Atomic.compare_and_set m old (Locked (r :: l)) then None
+        else block r (* failed CAS; retry *)
+  in
+  perform (Sched.Suspend block)
```

Comment:
Hmm... This is fine as an example, but note that this requires allocating a closure and perforning an effect (which also allocates an effect value).  For performance it would be significantly better to first attempt to take the lock without closure allocation and without performing effects.

Reply:
In Eio, the [mutex](https://ocaml-multicore.github.io/eio/eio/Eio/Mutex/index.html) has three states
1. Unlocked
2. Locked
3. [Poisoned](https://ocaml-multicore.github.io/eio/eio/Eio/Mutex/index.html#exception-Poisoned)

When poisoned, an attempt to lock the mutex raises an exception.  How would you implement such a feature with the `Suspend` effect?

May the `block` function passed as part of the effect value raise an exception and what happens if it does?